File: 474_ones_and_zeroes.java
class Solution {
    public int findMaxForm(String[] strs, int zeros, int ones) {
        int[][] dp = new int[zeros + 1][ones + 1];

        for (String str: strs) {
            int[] count = calZerosOnes(str, zeros, ones);
            for (int zero = zeros; zero >= count[0]; zero--) {
                for (int one = ones; one >= count[1]; one--) {
                    dp[zero][one] = Math.max(1 + dp[zero - count[0]][one - count[1]], dp[zero][one]);
                }
            }
        }

        return dp[zeros][ones];
    }

    private int[] calZerosOnes(String s, int zeros, int ones) {
        int[] c = new int[2];
        for (int i = 0; i < s.length(); i++) {
            c[s.charAt(i) - '0']++;
        }
        return c;
    }
}

File: python/474_ones_and_zeroes.py
class Solution:
    # Time Complexity: O(len(strs) * m * n)
    # Space Complexity: O(m * n)
    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]
        for s in strs:
            zero, one = self.countzeroesones(s)
            for i in range(m, zero - 1, -1):
                for j in range(n, one - 1, -1):
                    dp[i][j] = max(1 + dp[i - zero][j - one], dp[i][j])
        return dp[m][n]

    def countzeroesones(self, s: str):
        zero = one = 0
        for c in s:
            if c == '0':
                zero += 1
            elif c == '1':
                one += 1
        return zero, one
